### PR TITLE
Improve requests for resumable uploads.

### PR DESCRIPTION
**Pull request by coryan**
The initial implementation was incomplete. We also need a response
object and some tweaks to the request classes. This is part of the fixes
for #559.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1471)

### CODE REVIEW

File: google/cloud/storage/internal/object_requests.cc [hunk 1]
```diff
@@ -330,20 +330,81 @@ std::ostream& operator<<(std::ostream& os, ResumableUploadRequest const& r) {
...
+
+ResumableUploadResponse ResumableUploadResponse::FromHttpResponse(
+    HttpResponse&& response) {
+  ResumableUploadResponse result;
+  result.last_committed_byte = 0;
+  result.payload = std::move(response.payload);
+  if (response.headers.find("location") != response.headers.end()) {
+    result.upload_session_url = response.headers.find("location")->second;
+  }
+  if (response.headers.find("range") != response.headers.end()) {
```
Comment (houglum) [inline google/cloud/storage/internal/object_requests.cc]: This was a bit hard to follow until I pulled up the docs at https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload. Might be worth adding a comment here, e.g. `We expect a range header with its value formatted like "bytes=0-<number>", where <number> is the number of bytes received so far."  This explains that we always expect 0 as the first number in that range.

Actually, that being said... I think we could simplify lines 385 through 393 -- can we just search for the substring "bytes=0-" and return `result` if that's not found?

Comment (coryan) [reply google/cloud/storage/internal/object_requests.cc]: Thanks!  Fixed I think.

**Comment by coryan** [comment]
PTAL.
